feat: Multiple Hydro Rule Curves

nikolaredstork opened the pull request:
This PR implements the Multiple Rule Curves for Hydro Reservoirs as an optional new feature, RTE-I's CR24.

nikolaredstork committed:
Introduce new Time-Series that will represent maximum, minimum and average reservoir levels

nikolaredstork committed:
Intoduction to the new global parameter that will determine from which input files reservoir levels should be loaded

nikolaredstork committed:
Introduce reservoir levels objects into constructor

nikolaredstork committed:
Add comments

nikolaredstork committed:
Clang format and hydro reservoir level check

nikolaredstork committed:
Reservoir Levels load function

nikolaredstork committed:
Introduce new function for loading old reservoir levels files

nikolaredstork committed:
Add checks in solver

nikolaredstork committed:
Clang format

nikolaredstork committed:
Clang format

nikolaredstork committed:
Implement new objects to solver

nikolaredstork committed:
Bug Fix

nikolaredstork committed:
Refactoring and Unit tests

nikolaredstork committed:
Clang format

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr24

nikolaredstork committed:
Reservoir levels check revert

nikolaredstork committed:
Hydro Levels Check Fix

nikolaredstork committed:
Use getColum function

nikolaredstork committed:
Excluding unnecessary code

nikolaredstork committed:
Refactoring

nikolaredstork committed:
Save useScenarizedReservoirLevels value

nikolaredstork committed:
Clang format

nikolaredstork committed:
Change function name

nikolaredstork committed:
Additional check in reservoir levels check function

nikolaredstork committed:
Modification of load function

nikolaredstork committed:
Unit tests-Loading scenarized reservoir levels

nikolaredstork committed:
clang format

nikolaredstork committed:
Refactor hydro reservoir check

nikolaredstork committed:
Bug fix

nikolaredstork committed:
Unit tests for hydro inputs checker

nikolaredstork committed:
Clang format

nikolaredstork committed:
Exclud unnecessary code from unit tests

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr24

nikolaredstork committed:
Use Reservoir Levels enum

nikolaredstork committed:
Save new files. Added for testing

nikolaredstork committed:
Sonar cloud. To be removed on pushing this branch on RTE repo

nikolaredstork committed:
Add reservoir levels to tsCount function

nikolaredstork committed:
New class for hydro reservoir levels

nikolaredstork committed:
Use new objects

nikolaredstork committed:
Fix unit tests

nikolaredstork committed:
Excluding unnecessary code

nikolaredstork committed:
Refactoring

nikolaredstork committed:
Add comments

nikolaredstork committed:
Add copy function

nikolaredstork committed:
Unit test for load functions

nikolaredstork committed:
Adding more checks to unit test

nikolaredstork committed:
Fix folder creation in unit test

nikolaredstork committed:
Refactor loading functions

nikolaredstork committed:
Modify unit test

nikolaredstork committed:
Clang format

nikolaredstork committed:
UI bugfix

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Clang Format

nikolaredstork committed:
Move flag to compatibility section

nikolaredstork committed:
Naming convention

nikolaredstork committed:
Update user-guide

nikolaredstork committed:
Clang format

nikolaredstork committed:
Move flag to compatibility section

nikolaredstork committed:
Implement loader classes

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Exclude depraceted code

nikolaredstork committed:
Revert sonar workflow

nikolaredstork committed:
Possible fix

nikolaredstork committed:
Bugfix

nikolaredstork committed:
Move flag to compatibility part

nikolaredstork committed:
Clang format

nikolaredstork committed:
Clang format

nikolaredstork committed:
Hydro rule curves compatibility

nikolaredstork committed:
Clang format

nikolaredstork committed:
Documentation update

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Unit tests

nikolaredstork committed:
Clang format

nikolaredstork committed:
Add study version conversion

nikolaredstork committed:
Change function name

nikolaredstork committed:
Swithc places for initialization reasons

nikolaredstork committed:
Change name convention for object that communicate with GUI

nikolaredstork committed:
Clang Format

nikolaredstork committed:
Name convetion modification

nikolaredstork committed:
Naming convention second part

nikolaredstork committed:
Change name of files

nikolaredstork committed:
Move load from folder function to RuleCurvesLoader class

nikolaredstork committed:
Loader Service Class

nikolaredstork committed:
Add createRuleCurvesLoader in Service class

nikolaredstork committed:
Remove unnecessary code and modify unit tests

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Code smells

flomnes commented:
The coverage is a bit under 60%, can you please increase the coverage on hydrorulecurves.cpp ?
![image](https://github.com/user-attachments/assets/0ecb5b18-c4a0-4023-93c2-402516fead97)

flomnes commented on src/solver/hydro/management/HydroInputsChecker.cpp:
```diff
@@ -94,6 +98,50 @@ bool HydroInputsChecker::checkMinGeneration(uint year)
+              errorCollector_(area.name)
+                << "Reservoir levels Time-Series indexes in area: " << area.id
+                << " for year: " << year << " are not equal. Something went wrong!";
+          }
```
This check is both too strict and not useful
- Too strict : the user may want to have one column for either min, avg or max, this is usually valid since there is no ambiguity on what TS to chose since there is only one
- Not useful since `ruleCurves.min`, `ruleCurves.avg` and `ruleCurves.max` share the same `timeseriesNumbers`

Checks on the number of columns of co-scenarized TS are handled by `TimeSeriesNumbers::registerSeries` and `TimeSeriesNumbers::checkSeriesNumberOfColumnsConsistency`, so you don't need to do it yourself since you already added calls to `TimeSeriesNumbers::registerSeries`.

I suggest you remove this check.

nikolaredstork replied:
Done!

flomnes commented on src/tests/src/libs/antares/study/parts/hydro/test-hydro-series.cpp:
```diff
@@ -121,8 +168,8 @@ BOOST_FIXTURE_TEST_CASE(Testing_load_power_credits_matrices_equal_width, Fixture
 
-    fillTimeSeriesWithSpecialEnds(maxHourlyGenPower, 400.);
-    fillTimeSeriesWithSpecialEnds(maxHourlyPumpPower, 200.);
+    fillTimeSeriesWithSpecialEnds(maxHourlyGenPower, 401., 402.);
+    fillTimeSeriesWithSpecialEnds(maxHourlyPumpPower, 201., 202.);
```
Why did you have to change these values ?

nikolaredstork replied:
I wanted to reuse this function for filling up rule curves TS. Since `fillTimeSeriesWithSpecialEnds` was coded in the way that you need to parse float value and the function will add 1.0 to the parsed value and put it at the beginning of the TS and also it will add 2.0 to the same value and put it at the end of TS.

Rule Curves represent values between 0-1, so we couldn't reuse that function in previous implementation, so I've implemented in the way that I can now parse two values as arguments and put it at the beginning and at the end of the TS. 

Overall, only reason why I changed implementation of the function is to reuse it for filling up rule curve TS.

flomnes commented on src/libs/antares/study/version.cpp:
```diff
@@ -57,6 +57,12 @@ StudyVersion legacyVersionIntToVersion(unsigned version)
+    case 910:
+        return StudyVersion(9, 1);
+    case 900:
+        return StudyVersion(9, 0);
```
The format changed from version 9.0, we no longer accept 900, 910, 920, etc.

We did that to avoid having problems for versions 10.x

nikolaredstork replied:
We need it for our Antares Web deployment. I will exclude it.

flomnes commented on src/libs/antares/study/parts/hydro/hydrorulecurves.cpp:
```diff
@@ -0,0 +1,216 @@
+        if (enabledModeIsChanged)
+        {
+            JIT::enabled = true; // Back to the previous loading mode.
+        }
```
I have a feeling we could use the `StandardRuleCurveLoader` here (code duplication)
```cpp
if (!usedBySolver)  hydroRuleCurves = single;
```

nikolaredstork replied:
No we cannot, because this part of the code is for loading data when entering UI, same thing as for CR23. When we fully deprecate old UI, this part of the code can be excluded. Also when we fully deprecate old UI, `standardRuleCurvesGUI` class member can be also excluded and just use local variable to load data from standard rule curve file.

nikolaredstork committed:
Use unsigned integer literal

nikolaredstork committed:
Exclude unnecessary checks

nikolaredstork committed:
Additional unit tests

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Clang format

nikolaredstork committed:
Post merging clean

nikolaredstork committed:
Add test

nikolaredstork committed:
Reduce number of conditions

nikolaredstork committed:
Clang format

nikolaredstork committed:
Add test for UI data loading

nikolaredstork committed:
Revert change for legacy versioning

nikolaredstork committed:
Revert changes

nikolaredstork committed:
Monthly unit test as on develop

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24 for custom deployment

dusanparipovic committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

dusanparipovic committed:
style(clang-format): add clang format to monthly.cpp

dusanparipovic committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

dusanparipovic committed:
Synchronize new object into exsisting function

nikolaredstork committed:
initial refactor - UI excluded

nikolaredstork committed:
Clang Format

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Exclude unnecessary function arguments

nikolaredstork committed:
Code smell

guilpier-code commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -235,6 +235,40 @@ bool StringToCompatibilityHydroPmax(Parameters::Compatibility::HydroPmax& mode,
+    if (text.empty())
+    {
+        return false;
+    }
```
```c++
if (text.empty())
{
    return false;
}
```
is useless : if we remove this block, behavior does not change.

guilpier-code replied:
Same for **StringToCompatibilityHydroPmax(...)**

guilpier-code replied:
Done

nikolaredstork replied:
This was proposed in previous implementation, if refactoring is necessary, it should be done in different PR.

guilpier-code replied:
Done in this [(corrective) PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3310)

guilpier-code commented on src/libs/antares/study/include/antares/study/parameters.h:
```diff
@@ -363,6 +363,12 @@ class Parameters final
             Hourly
         };
         HydroPmax hydroPmax = HydroPmax::Daily;
+        enum class HydroRuleCurves
+        {
+            Single,
+            Scenarized
+        };
+        HydroRuleCurves hydroRuleCurves = HydroRuleCurves::Single;
     };
```
_**Personal remark**_ (not mandatory to apply it)

I personally don't like struct / class defined inside other struct / class declaration :  it makes surrounding class declaration bigger than it should.
Class Parameters is already very big (and should be split, by the way) and we inflate it we struct declaration.
A class declaration is a kind of summary of the class,  it should be as short as possible, so we can see at a glance what it does.
If, like me, you don't like big classes, help moving inner struct declarations outside the surrounding class.

nikolaredstork replied:
This was proposed in previous implementation, if refactoring is necessary, it should be done in different PR.

guilpier-code commented on src/libs/antares/study/include/antares/study/parameters.h:
```diff
@@ -363,6 +363,12 @@ class Parameters final
             Hourly
         };
         HydroPmax hydroPmax = HydroPmax::Daily;
+        enum class HydroRuleCurves
+        {
+            Single,
+            Scenarized
+        };
+        HydroRuleCurves hydroRuleCurves = HydroRuleCurves::Single;
```
If I understand correctly, either rules curves are scenarized, or it's not.
So why not a simple boolean for this information  ?
```c++
bool hydroRuleCurvesScenarized = false; // Default value
```

nikolaredstork replied:
This was proposed in previous implementation, if refactoring is necessary, it should be done in different PR.

guilpier-code commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,179 @@
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#ifndef __ANTARES_LIBS_STUDY_PARTS_HYDRO_RESERVOIR_LEVELS_H__
```
Please replace : 
```c++
#ifndef __ANTARES_LIBS_STUDY_PARTS_HYDRO_RESERVOIR_LEVELS_H__
#define __ANTARES_LIBS_STUDY_PARTS_HYDRO_RESERVOIR_LEVELS_H__
...
#endif
```
with just : 
```c++
#pragma once
```

guilpier-code replied:
Done in this [(corrective) PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3310)

guilpier-code commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,179 @@
+#include <antares/study/parameters.h>
+#include <antares/study/version.h>
+
+#include "../../fwd.h"
```
We don't need to include these headers :
```c++
#include <antares/array/matrix.h>
#include <antares/study/version.h>
#include "../../fwd.h"
```
It slows down compilation time and makes code less simple than it could be.
Please remove them.

guilpier-code replied:
Done in this [(corrective) PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3310)

guilpier-code commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,179 @@
+
+#include "../../fwd.h"
+
+namespace Antares
```
Please concatenate namespaces : 
```c++
namespace Antares::Data
{
   ...
}
```

guilpier-code replied:
Done in this [(corrective) PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3310)

guilpier-code commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,179 @@
+public:
+    enum
+    {
+        //! The minimum value
```
Comments here bring nothing more than the **enum** items, we should remove them

guilpier-code replied:
Done

guilpier-code replied:
Done in this [(corrective) PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3310)

guilpier-code commented on src/libs/antares/study/parts/hydro/hydrorulecurves.cpp:
```diff
@@ -0,0 +1,146 @@
+
+#include <algorithm>
+
+#include <yuni/yuni.h>
```
It seems we don't need all of these header inclusions : 
```c++
#include <algorithm>

#include <yuni/yuni.h>
#include <yuni/io/file.h>
#include <antares/exception/LoadingError.hpp>
#include <antares/inifile/inifile.h>
#include <antares/logs/logs.h>
#include "antares/study/study.h"
```
Please remove them if I'm right

guilpier-code replied:
Done

guilpier-code replied:
Done in this [(corrective) PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3310)

guilpier-code commented on src/libs/antares/study/parts/hydro/hydrorulecurves.cpp:
```diff
@@ -0,0 +1,146 @@
+#include <antares/study/parts/hydro/hydrorulecurves.h>
+#include "antares/study/study.h"
+
+namespace fs = std::filesystem;
```
You specified : 
```c++
namespace fs = std::filesystem;
```
and yet barely use it : there are still many **std::filesystem::** in this source file.

guilpier-code commented on src/libs/antares/study/parts/hydro/hydrorulecurves.cpp:
```diff
@@ -0,0 +1,146 @@
+  const std::filesystem::path& folder,
+  Parameters::Compatibility::HydroRuleCurves hydroRuleCurves)
+{
+    bool ret = true;
```
```c++
bool ret = true;
```
is useless here.
```c++
...
return loader->load();
```
is enough.

guilpier-code replied:
Done

guilpier-code commented on src/libs/antares/study/parts/hydro/hydrorulecurves.cpp:
```diff
@@ -0,0 +1,146 @@
+    bool ret = true;
+
+    fs::path filePath = path / "maxDailyReservoirLevels.txt";
+    ret &= max_.timeSeries.loadFromCSVFile(filePath.string(), 1, DAYS_PER_YEAR, &fileContent);
```
Here operator `&=` is used.
This operator seems more associated to **bitwise AND** operator than to **logical AND** operator.
If a `&&=` operator existed, we would like to use it here, but it does not seem to exist.
So I guess `ret = expr && ret` should be used instead.

guilpier-code commented on src/libs/antares/study/parts/hydro/hydrorulecurves.cpp:
```diff
@@ -0,0 +1,146 @@
+bool ScenarizedRuleCurvesLoader::load()
+{
+    const std::filesystem::path path = baseFolder_ / "series" / areaID_;
+    Matrix<>::BufferType fileContent;
```
**_Out of this PR : simple observation_**
```c++
Matrix<>::BufferType fileContent;
```
OMG : this variable is requested as argument of the loading function, but is never used in the calling scope. 
Why then pass it as argument ? 
It seems it does not need to be passed, and function signature should be simplified. 
To be looked at, some day (very soon), when legacy GUI is removed as planned.

nikolaredstork replied:
This `Matrix<>::BufferType fileContent` is USED in all loading functions, I've just went with your approach.

For example:

<img width="858" height="301" alt="image" src="https://github.com/user-attachments/assets/c33873e5-4313-45f4-81b9-1e8e089904c5" />

guilpier-code replied:
Yes, that's why my remark was tagged "out of this PR".
But still the question remains : why do we pass an argument to a function, here an empty string buffer, if it's not to do something with it after the call ?

guilpier-code replied:
Let me be more clear : we need the string buffer in the body of the function, internally. That I don't deny.
But it does not seem to be used ever in the client side (the function caller). 
So this would tend to prove that this buffer should be defined as a local variable inside the loading function, and disappear from the list of its argument.
So do we ever do something with the argument on the client side ? 
I admit I did not check.

nikolaredstork replied:
Totally agree. 
Function `internalLoadCSVFile` from **matrix.hxx** file should be reviewed:
```
template<class T, class ReadWriteT>
bool Matrix<T, ReadWriteT>::internalLoadCSVFile(const AnyString& filename,
                                                uint minWidth,
                                                uint maxHeight,
                                                uint options,
                                                BufferType* buffer) 
```

guilpier-code committed:
CR 24 : trial for code improvements (#3310)

This PR is based on the [PR about multiple hydro rule
curves](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2829)
(CR 24).
It suggests quick and easy code improvements.

---------

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

flomnes committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

flomnes committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
fix tests

nikolaredstork committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

nikolaredstork committed:
Clang Format

flomnes committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

flomnes commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,139 @@
+    }
+
+private:
+    bool load() final;
```
```suggestion
    bool load() override final;
```

flomnes commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,139 @@
+                               TimeSeries& min):
+        RuleCurvesLoader(baseFolder, areaID, max, avg, min)
+    {
+    }
```
Simplification - use constructor delegation
```suggestion
    using RuleCurvesLoader::RuleCurvesLoader;
```

flomnes commented on src/libs/antares/study/include/antares/study/parts/hydro/hydrorulecurves.h:
```diff
@@ -0,0 +1,139 @@
+        RuleCurvesLoader(baseFolder, areaID, max, avg, min)
+
+    {
+    }
```
Use constructor delegation - see below

flomnes commented on docs/user-guide/04-migration-guides.md:
```diff
@@ -7,6 +7,24 @@ lower the costs of changing existing interfaces, both GUI and scripts.
 
 ### Input
 
+#### New optional feature and compatibility flag for scenarized hydro reservoir levels
```
planned for 10.1.0

flomnes commented on src/libs/antares/study/include/antares/study/parameters.h:
```diff
@@ -468,6 +474,10 @@ bool StringToSimulationMode(SimulationMode& mode, Yuni::CString<20, false> text)
 bool StringToCompatibilityHydroPmax(Parameters::Compatibility::HydroPmax&, const std::string& text);
 
+const char* CompatibilityHydroRuleCurvesToCString(
+  const Parameters::Compatibility::HydroRuleCurves mode);
```
unnecessary (no more save functions)

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -202,6 +198,36 @@ bool StringToCompatibilityHydroPmax(Parameters::Compatibility::HydroPmax& mode,
+    default:
+        return "Unknown";
+    }
+}
```
unnecessary (no more save functions)

flomnes commented on src/solver/hydro/management/daily.cpp:
```diff
@@ -218,7 +218,8 @@ inline void HydroManagement::prepareDailyOptimalGenerations(
 
-    auto& lowLevel = area.hydro.reservoirLevel[Data::PartHydro::minimum];
+    uint32_t seriesIndex = area.hydro.series->ruleCurves.min.getSeriesIndex(y);
+    const auto& lowLevel = area.hydro.series->ruleCurves.min.timeSeries[seriesIndex];
```
```suggestion
    const auto* lowLevel = area.hydro.series->ruleCurves.min.timeSeries[seriesIndex];
```

flomnes commented on src/solver/hydro/management/HydroInputsChecker.cpp:
```diff
@@ -77,6 +81,35 @@ bool HydroInputsChecker::checkMinGeneration(uint year)
+      {
+          const auto& minRuleCurves = area.hydro.series->ruleCurves.min.getColumn(year);
+          const auto& avgRuleCurves = area.hydro.series->ruleCurves.avg.getColumn(year);
+          const auto& maxRuleCurves = area.hydro.series->ruleCurves.max.getColumn(year);
```
```suggestion
          const auto* minRuleCurves = area.hydro.series->ruleCurves.min.getColumn(year);
          const auto* avgRuleCurves = area.hydro.series->ruleCurves.avg.getColumn(year);
          const auto* maxRuleCurves = area.hydro.series->ruleCurves.max.getColumn(year);
```

flomnes commented on src/solver/hydro/management/monthly.cpp:
```diff
@@ -256,12 +255,12 @@ void HydroManagement::prepareMonthlyOptimalGenerations(
           else
           {
-              auto& reservoirLevel = area.hydro.reservoirLevel[Data::PartHydro::average];
+              const auto& ruleCurves = area.hydro.series->ruleCurves.avg.getColumn(y);
```
```suggestion
              const auto* ruleCurves = area.hydro.series->ruleCurves.avg.getColumn(y);
```

flomnes commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -711,7 +711,9 @@ void SIM_RenseignementProblemeHebdo(const Study& study,
                                                                 .dayYear;
-                                auto& minLvl = area.hydro.reservoirLevel[Data::PartHydro::minimum];
+
+                                const auto& minLvl = area.hydro.series->ruleCurves.min.getColumn(
```
```suggestion
                                const auto* minLvl = area.hydro.series->ruleCurves.min.getColumn(
```

flomnes commented on src/solver/simulation/sim_calcul_economique.cpp:
```diff
@@ -856,8 +858,9 @@ void SIM_RenseignementProblemeHebdo(const Study& study,
-                                    auto& maxLvl = area.hydro
-                                                     .reservoirLevel[Data::PartHydro::maximum];
+
+                                    const auto& maxLvl = area.hydro.series->ruleCurves.max
```
```suggestion
                                    const auto* maxLvl = area.hydro.series->ruleCurves.max
```

flomnes committed:
Apply hydro rule-curve review fixes

flomnes committed:
format

flomnes committed:
Cleaning

flomnes committed:
auto*

flomnes committed:
Move impl classes to .cpp

dusanparipovic commented:
Hi @flomnes I'll take a look asap.

flomnes commented:
@dusanparipovic I think I fixed most of the issues, the PR can finally be merged

dusanparipovic commented:
> @dusanparipovic I think I fixed most of the issues, the PR can finally be merged

If you want anything additional ping me, I'm looking for contribution on simulator :) @flomnes

JasonMarechal25 committed:
Merge branch 'develop' into feature/multiple-rule-curve-cr-24

Commit:
doc

Commit:
Fix build

Commit:
Fix btest
